azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/message_queue.py

```diff
@@ -0,0 +1,138 @@
+        self._on_disable = on_disable
+
+    def enable(self):
+        self._enabled = True
+        self._on_enable()
+
+    def disable(self):
+        self._enabled = False
+        self._on_disable()
+
+    @property
+    def enabled(self):
+        return self._enabled
+
+    @enabled.setter
+    def enabled(self, v):
+        if v:
+            self.enable()
+        else:
+            self.disable()
+
+
+class MessageQueueManager(object):
+    def __init__(self, transport):
+        self.c2d_message_queue = MessageQueue(
+            self._c2d_message_queue_on_enable, self._c2d_message_queue_on_disable
+        )  # not yet enabled
+        self.input_message_queues = {}
+        self._transport = transport
+
+    def _input_message_queue_on_enable(self):
+        self._enable_feature("input")
+        # Maybe check if input is enabled first?
+
+    def _c2d_message_queue_on_enable(self):
+        self._enable_feature("c2d")
+
+    def _input_message_queue_on_disable(self):
+        # Disable input messages as a feature if all input queues are disabled
+        if all(not queue.enabled for queue in self.input_message_queues.values()):
+            self._disable_feature("input")
+
+    def _c2d_message_queue_on_disable(self):
+        self._disable_feature("c2d")
+
+    def _enable_feature(self, feature_name):
+        """
+        To enable a specific feature on the internal client.Some of the features that can be enabled
+        are "receiving input messages" and "receiving cloud to device messages"
+        :param feature_name: The specific feature which needs to be enabled. Feature names can be:-
+        "input" , "c2d" etc.
+        "inputs" and on the attribute "on_c2d_message" for "c2d"
+        :raises ValueError if feature_name is not amongst "input" or "c2d"
+        """
+        enable_complete = Event()
+
+        def callback():
+            enable_complete.set()
+
+        if feature_name == "input":
+            self._transport.enable_input_messages(callback)
+        elif feature_name == "c2d":
+            self._transport.enable_c2d_messages(callback)
+        else:
+            logger.error("Feature name has not been defined.Feature names can be 'input' or 'c2d'")
+            raise ValueError("Feature names can be only among 'input' or 'c2d'")
+
+        enable_complete.wait()
+
+    def _disable_feature(self, feature_name):
+        disable_complete = Event()
+
+        def callback():
+            disable_complete.set()
+
+        if feature_name == "input":
+            self._transport.disable_input_messages(callback)
+        elif feature_name == "c2d":
+            self._transport.disable_c2d_messages(callback)
+        else:
+            logger.error("Feature name has not been defined.Feature names can be 'input' or 'c2d'")
+            raise ValueError("Feature names can be only among 'input' or 'c2d'")
+
+        disable_complete.wait()
+
+    def get_input_message_queue(self, input_name):
```

Comment:
>get_input_message_queue [](start = 8, length = 23)

if you wanted to commit fully to the use of strings/enums, shouldn't this be get_feature()?

Reply:
I'm not sure what you mean by this. This method returns a specified input message queue, not a feature.

---
In reply to: [258258677](https://github.com/Azure/azure-iot-sdk-python-preview/pull/49#discussion_r258258677) [](ancestors = 258258677)

Reply:
You have get_input_message_queue and get_c2d_message_queue (ie one function per feature) however you have "enable_feature" and "disable_feature" where I'm guessing "feature_name" could be either input or c2d. that's where the API is asymetrical.

---
In reply to: [259172314](https://github.com/Azure/azure-iot-sdk-python-preview/pull/49#discussion_r259172314) [](ancestors = 259172314,258258677)